Common/include/matrix_structure.hpp

```diff
@@ -280,127 +248,90 @@ class CSysMatrix {
   void SubtractBlock_ILUMatrix(unsigned long block_i, unsigned long block_j, ScalarType *val_block);
   
   /*!
-   * \brief Adds the specified value to the diagonal of the (i, i) subblock
-   *        of the matrix-by-blocks structure.
-   * \param[in] block_i - Index of the block in the matrix-by-blocks structure.
-   * \param[in] val_matrix - Value to add to the diagonal elements of A(i, i).
-   */
-  template<class OtherType>
-  void AddVal2Diag(unsigned long block_i, OtherType val_matrix);
-  
-  /*!
-   * \brief Sets the specified value to the diagonal of the (i, i) subblock
-   *        of the matrix-by-blocks structure.
-   * \param[in] block_i - Index of the block in the matrix-by-blocks structure.
-   * \param[in] val_matrix - Value to add to the diagonal elements of A(i, i).
+   * \brief Performs the product of i-th row of the upper part of a sparse matrix by a vector.
+   * \param[in] vec - Vector to be multiplied by the upper part of the sparse matrix A.
+   * \param[in] row_i - Row of the matrix to be multiplied by vector vec.
+   * \return prod Result of the product U(A)*vec (stored at *prod_row_vector).
    */
-  template<class OtherType>
-  void SetVal2Diag(unsigned long block_i, OtherType val_matrix);
+  void UpperProduct(const CSysVector<ScalarType> & vec, unsigned long row_i);
   
   /*!
-   * \brief Calculates the matrix-vector product
-   * \param[in] matrix
-   * \param[in] vector
-   * \param[out] product
+   * \brief Performs the product of i-th row of the lower part of a sparse matrix by a vector.
```

Comment:
Thanks for attention to detail in the comments

Reply:
Borderline ocd... but oh well.